### DISCUSSION TITLE
How to keep track of variable scope?

### DISCUSSION THREAD

**Discussion by FnControlOption** [lib-ruby-parser/lib-ruby-parser#68] [Q&A]
Currently experimenting with replacing mruby's parser with lib-ruby-parser, and I've seem to run into a bit of a problem... mruby's parser keeps track of which local variables are in scope at any point in the AST, while lib-ruby-parser doesn't appear to do the same, at least not in a way that can be accessed externally. I'm thinking an alternative might be to manually traverse the AST after having already parsed it and scanning for local variables that way, but that feels inefficient. Any ideas?

P.S. a C binding for `DecodedInput#line_col_for_pos` would be nice 😉

**Comment by iliabylich** [comment]
You are right, currently there's no way to get all locals declared in a scope.

So if we add it to `lib-ruby-parser` I assume it should be accessible on the following list of nodes:

1. `def`
2. `defs`
3. `class` (they do introduce new "closure")
4. `module` (same)
5. `begin` (for both blocks and top-level scopes)

Which makes API complicated and it doesn't feel like something that must be available in a general-purpose-parser. Should there also be a similar API for all constant declarations? Or for all method definitions? I'd prefer to avoid them.

Do you know how mruby/MRI pass this information to `compile.c`? I haven't checked it but I guess they both handle it on-the-fly, i.e. when they compile `lvasgn('a', int('1'))` they "record" that there's a local variable `a` and push it into some stack of sets of variable. I hope it doesn't sound confusing.

> P.S. a C binding for DecodedInput#line_col_for_pos would be nice 😉

Do you want to send a PR? If not I will do it myself later tomorrow.

**Reply by iliabylich** [reply]
@FnControlOption I've released [v4.0.4+ruby-3.1.1](https://github.com/lib-ruby-parser/c-bindings/releases/tag/v4.0.4%2Bruby-3.1.1) with [`LIB_RUBY_PARSER_line_col_for_pos`](https://lib-ruby-parser.github.io/c-bindings/group__decoded__input.html#gaf2de9fa1bd7c8aabc8e673f941bed34d) function, PTAL. Do you use pre-compiled build by the way? Or do you build it manually?
